### DISCUSSION TITLE
How to access incoming requests at Uvicorn level in a FastAPI app?

### DISCUSSION THREAD

**Discussion by raghavsikaria** [Kludex/uvicorn#1314] [Q&A] [answered]
Hey, community!

I have a FastAPI application that I have deployed using the UVICORN server. My requirement is such that I have to log every incoming request at the UVICORN level and also decode JWT Token that the header of incoming request contains, log that too, at the UVICORN level.

I have been trying to understand the UVICORN library and understanding the code. So far, this is what I have found:

https://github.com/encode/uvicorn/blob/df8b49f6779b9de508416ad3823bac868bb8efcf/uvicorn/protocols/http/httptools_impl.py#L128-L138

With some hit & trial, I found that the data parameter this function receives contains all the incoming request payload as a byte string.
But I am still not able to locate the exact place where I can find the incoming request and its data.

Where can I find all incoming requests and go about my logging? I don't want to monkey-patch the UVICORN code, I am looking for a sane way to achieve the purpose.

**Comment by Kludex** [comment, marked as answer]
The best way to achieve that, right now, is to create a middleware. Check https://github.com/Kludex/asgi-logger internal code for more details.

**Reply by raghavsikaria** [reply]
hey @Kludex thanks for the prompt reply! I browsed through the asgi-logger project. It seems that this middleware is added at the FastAPI app level. However, I want to add middleware at the uvicorn level, how to work with that?

**Reply by Kludex** [reply]
It's not possible right now, and it's not an issue to do it with a middleware (the way in the link above).

Also, it's not FastAPI specific. It can be used with any ASGI web framework, jfyk.

**Reply by raghavsikaria** [reply]
oh okay. Thanks. 
Also, here is my motivation and some context: we noticed in our Django application served via Apache server, that there are some requests which reach Apache server but for some various reasons are not able to make it to Django application. So we log requests at the Apache level too. To combat the possibility of the same issue in our FASTAPI app, we thought of logging requests at the UVICORN level. Could you suggest to me any other way to monitor these requests?

**Reply by Kludex** [reply]
I don't have another way.

**Reply by raghavsikaria** [reply]
Oh okay, thanks!
